File: engine/camera.py
"""Dynamic camera that tracks all active fighters."""

from __future__ import annotations
from typing import List, TYPE_CHECKING

import pygame

from settings import SCREEN_WIDTH, SCREEN_HEIGHT

if TYPE_CHECKING:
    from entities.fighter import Fighter
    from stages.stage import Stage


class Camera:
    """Smoothly follows the midpoint of all fighters with dynamic zoom."""

    PADDING = 200       # extra pixels around fighters
    MIN_ZOOM = 0.35
    MAX_ZOOM = 1.0
    LERP_SPEED = 4.0    # how fast camera catches up (per second)

    def __init__(self):
        self.x: float = 0.0
        self.y: float = 0.0
        self.zoom: float = 1.0
        self._target_x: float = 0.0
        self._target_y: float = 0.0
        self._target_zoom: float = 1.0

    # ------------------------------------------------------------------
    def update(self, fighters: List[Fighter], stage: Stage, dt: float) -> None:
        alive = [f for f in fighters if f.stocks > 0]
        if not alive:
            return

        # Bounding box of all fighters
        min_x = min(f.x for f in alive)
        max_x = max(f.x + f.width for f in alive)
        min_y = min(f.y for f in alive)
        max_y = max(f.y + f.height for f in alive)

        # Target center
        self._target_x = (min_x + max_x) / 2
        self._target_y = (min_y + max_y) / 2

        # Zoom to fit all fighters
        span_x = (max_x - min_x) + self.PADDING * 2
        span_y = (max_y - min_y) + self.PADDING * 2
        zoom_x = SCREEN_WIDTH / max(span_x, 1)
        zoom_y = SCREEN_HEIGHT / max(span_y, 1)
        self._target_zoom = max(self.MIN_ZOOM, min(self.MAX_ZOOM, min(zoom_x, zoom_y)))

        # Lerp towards target
        t = min(1.0, self.LERP_SPEED * dt)
        self.x += (self._target_x - self.x) * t
        self.y += (self._target_y - self.y) * t
        self.zoom += (self._target_zoom - self.zoom) * t

    # ------------------------------------------------------------------
    def world_to_screen(self, wx: float, wy: float) -> tuple[float, float]:
        """Convert world coordinates to screen coordinates."""
        sx = (wx - self.x) * self.zoom + SCREEN_WIDTH / 2
        sy = (wy - self.y) * self.zoom + SCREEN_HEIGHT / 2
        return sx, sy

    def screen_to_world(self, sx: float, sy: float) -> tuple[float, float]:
        wx = (sx - SCREEN_WIDTH / 2) / self.zoom + self.x
        wy = (sy - SCREEN_HEIGHT / 2) / self.zoom + self.y
        return wx, wy

    def apply_surface(self, surface: pygame.Surface, world_x: float, world_y: float,
                      dest: pygame.Surface) -> None:
        """Blit a surface at world coords onto the destination surface."""
        sx, sy = self.world_to_screen(world_x, world_y)
        w = int(surface.get_width() * self.zoom)
        h = int(surface.get_height() * self.zoom)
        if w < 1 or h < 1:
            return
        scaled = pygame.transform.scale(surface, (w, h))
        dest.blit(scaled, (sx, sy))

    def get_visible_rect(self) -> pygame.Rect:
        """Return world-space rect of what's currently visible."""
        half_w = (SCREEN_WIDTH / 2) / self.zoom
        half_h = (SCREEN_HEIGHT / 2) / self.zoom
        return pygame.Rect(
            int(self.x - half_w), int(self.y - half_h),
            int(half_w * 2), int(half_h * 2),
        )

File: js/Camera.js
/**
 * Camera.js — Dynamic camera with distance-based zoom scaling.
 *
 * ══════════════════════════════════════════════════════════════════
 *  ZOOM FORMULA
 * ══════════════════════════════════════════════════════════════════
 *  span   = max player spread + PADDING * 2
 *  zoomX  = canvasW / span.x
 *  zoomY  = canvasH / span.y
 *  target = clamp(min(zoomX, zoomY), MIN_ZOOM, MAX_ZOOM)
 *
 *  As fighters spread apart the camera smoothly zooms out.
 *  As they cluster it zooms back in (up to MAX_ZOOM).
 *  Smooth interpolation via exponential lerp (LERP factor).
 *
 *  Camera can be optionally bounded by stage.cameraBounds to
 *  prevent scrolling past the edge of large maps.
 * ══════════════════════════════════════════════════════════════════
 */
(function() {
const S = SMASH.Settings;

class Camera {
    constructor() {
        this.x    = 0;      // world-space center focus x
        this.y    = 0;      // world-space center focus y
        this.zoom = 1;      // current zoom factor

        this._tx  = 0;      // target x
        this._ty  = 0;      // target y
        this._tz  = 1;      // target zoom
    }

    // ── Tuning constants ─────────────────────────────────────────
    static PADDING      = 250;     // extra space around player spread
    static MIN_ZOOM     = 0.22;    // furthest zoom-out (large maps)
    static MAX_ZOOM     = 1.15;    // closest zoom-in
    static DEFAULT_ZOOM = 0.7;     // zoom when only one player
    static LERP_POS     = 4.5;     // position smoothing speed
    static LERP_ZOOM    = 3.0;     // zoom smoothing speed
    static LEAD_FACTOR  = 0.08;    // look-ahead based on avg velocity

    update(fighters, stage, dt) {
        const alive = fighters.filter(f => f.stocks > 0);
        if (!alive.length) return;

        // ── Bounding box of all alive fighters ───────────────────
        let minX = Infinity,  maxX = -Infinity;
        let minY = Infinity,  maxY = -Infinity;
        let avgVx = 0, avgVy = 0;

        for (const f of alive) {
            const fx1 = f.x;
            const fx2 = f.x + f.width;
            const fy1 = f.y;
            const fy2 = f.y + f.height;
            if (fx1 < minX) minX = fx1;
            if (fx2 > maxX) maxX = fx2;
            if (fy1 < minY) minY = fy1;
            if (fy2 > maxY) maxY = fy2;
            avgVx += f.vx;
            avgVy += f.vy;
        }
        avgVx /= alive.length;
        avgVy /= alive.length;

        // ── Target position (center of bounding box + lead) ─────
        const cx = (minX + maxX) / 2;
        const cy = (minY + maxY) / 2;
        this._tx = cx + avgVx * Camera.LEAD_FACTOR;
        this._ty = cy + avgVy * Camera.LEAD_FACTOR;

        // ── Target zoom (based on spread) ────────────────────────
        const spanX = (maxX - minX) + Camera.PADDING * 2;
        const spanY = (maxY - minY) + Camera.PADDING * 2;

        if (alive.length === 1) {
            // Single fighter: comfortable default zoom
            this._tz = Camera.DEFAULT_ZOOM;
        } else {
            const zx = S.W / Math.max(spanX, 1);
            const zy = S.H / Math.max(spanY, 1);
            this._tz = Math.max(Camera.MIN_ZOOM,
                       Math.min(Camera.MAX_ZOOM, Math.min(zx, zy)));
        }

        // ── Smooth interpolation ─────────────────────────────────
        const tPos  = Math.min(1, Camera.LERP_POS * dt);
        const tZoom = Math.min(1, Camera.LERP_ZOOM * dt);

        this.x    += (this._tx - this.x) * tPos;
        this.y    += (this._ty - this.y) * tPos;
        this.zoom += (this._tz - this.zoom) * tZoom;

        // ── Camera bounds (optional stage constraint) ────────────
        if (stage && stage.cameraBounds) {
            const b = stage.cameraBounds;
            // Compute the viewable world-space half-dimensions at current zoom
            const halfW = (S.W / 2) / this.zoom;
            const halfH = (S.H / 2) / this.zoom;

            // Clamp so the camera viewport doesn't show beyond bounds
            if (b.minX !== undefined) this.x = Math.max(b.minX + halfW, this.x);
            if (b.maxX !== undefined) this.x = Math.min(b.maxX - halfW, this.x);
            if (b.minY !== undefined) this.y = Math.max(b.minY + halfH, this.y);
            if (b.maxY !== undefined) this.y = Math.min(b.maxY - halfH, this.y);
        }
    }

    /** World → screen X */
    wtsx(wx) { return (wx - this.x) * this.zoom + S.W / 2; }
    /** World → screen Y */
    wtsy(wy) { return (wy - this.y) * this.zoom + S.H / 2; }

    /** Screen → world X */
    stwx(sx) { return (sx - S.W / 2) / this.zoom + this.x; }
    /** Screen → world Y */
    stwy(sy) { return (sy - S.H / 2) / this.zoom + this.y; }

    /** Get world-space viewport rect */
    getViewport() {
        const halfW = (S.W / 2) / this.zoom;
        const halfH = (S.H / 2) / this.zoom;
        return {
            x: this.x - halfW,
            y: this.y - halfH,
            w: halfW * 2,
            h: halfH * 2,
        };
    }
}

SMASH.Camera = Camera;
})();
